--- lparser.c
+++ lparser.c
@@ -8409,7 +8409,11 @@
             if (strcmp(name, "_if") == 0 || strcmp(name, "asmif") == 0) {
               nest_level++;
             }
             else if (strcmp(name, "_endif") == 0 || strcmp(name, "asmend") == 0) {
-              nest_level--;
+              if (nest_level == 1) {
+                luaX_next(ls);
+                nest_level = 0;
+                break;
+              } else { nest_level--; }
             }
             else if (nest_level == 1 && (strcmp(name, "_else") == 0 || strcmp(name, "asmelse") == 0)) {
